implement kubectl procelain csr commands

mikedanese opened the pull request:
cc @gtank

ref #30163

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30237)

<!-- Reviewable:end -->

pwittrock commented on pkg/kubectl/cmd/alpha.go:
```diff
@@ -0,0 +1,40 @@
+func NewCmdAlpha(f *util.Factory, out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "alpha SUBCOMMAND",
+		Short: "Alpha versions of gcloud commands.",
```
Text seems wrong

pwittrock commented:
Tests?

pwittrock commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,166 @@
+	cmd := &cobra.Command{
+		Use:   "certificate SUBCOMMAND",
+		Short: "Modify certificate resources.",
+		Long:  ``,
```
Is this intentionally left empty?

pwittrock commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,166 @@
+	cmd := &cobra.Command{
+		Use:   "approve SUBCOMMAND",
+		Short: "approve a CSR.",
+		Long:  ``,
```
Same comment here

pwittrock commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,166 @@
+		Use:   "certificate SUBCOMMAND",
+		Short: "Modify certificate resources.",
+		Long:  ``,
+		Run: func(cmd *cobra.Command, args []string) {
```
Add an example

pwittrock commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,166 @@
+		csr.Status.Conditions = append(csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
+			Type:           certificates.CertificateDenied,
+			Reason:         "KubectlDeny",
+			Message:        "This csr was approved by kubectl alpha certificate deny",
```
should this say that the csr was denied?

smarterclayton replied:
Saying the CLI command that denied it is a bit tacky.  I suspect we should support "--message" with a default that is "Denied by administrator (%s)" and substitute the full path from cmd.CommandPath() rather than hardcoding.

smarterclayton replied:
Tacky as in its passive voice - the administrator denied the request, not the CLI.  Always identify the who.

mikedanese commented:
uhh. this PR is not for the master branch??

kargakis commented:
cc: @kubernetes/kubectl

kargakis commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,168 @@
+		if alreadyApproved {
+			return csr, "approved"
+		}
+		csr.Status.Conditions = append(csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
```
It's strange that a client command is updating the status of the csr directly. Shouldn't this be handled by the controller?

mikedanese replied:
Yes it is. This causes the controller to issue a certificate for the CSR. The design talked about a an approval process (basically a separate controller) that lives outside k8s and approves CSRs that steer then signed by the signing controller which lives in the controller manager. This command enables the "human approval controller".

smarterclayton replied:
I think of this as an admin command (like node operations).  We have other actions already that are quasi controller behavior (drain) that humans also run.

In the long description I would recommend explaining how to use this command in a real (at least quasi real) workflow.

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,168 @@
+		}
+		csr := info.Object.(*certificates.CertificateSigningRequest)
+		csr, verb := modify(csr)
+		err = info.Client.Put().
```
Should be using generated client - this directly exposes raw actions and the client abstracts that intentionally.  Also client is versioned.

mikedanese replied:
Fixed.

mikedanese committed:
implement kubectl procelain csr commands

mikedanese committed:
autogenerated

mikedanese commented:
@pwittrock I've addressed comments. Can you find someone to assign this to or finish review?

@zmerlynn @roberthbailey can one of you review the first commit? It's bash :)

mikedanese commented:
hmm, I broke test-cmd.sh, but other stuff is ready for a new review.

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,174 @@
+
+SECURITY NOTICE: Depending on the requested attributes, the issued certificate
+can potentially grant a requester access to cluster resources or to authenticate
+as a requested identity. This can be very useful but the security reprecussion
```
I feel like this warning isn't warning enough :)

Second sentence:  "Before approving a CSR, ensure you understand what the signed certificate can do."  or similar

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,174 @@
+func NewCmdCertificateApprove(f cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := CertificateOptions{}
+	cmd := &cobra.Command{
+		Use:   "approve CSR",
```
This is confusing, should be "approve CSR_NAME" or similar.

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,174 @@
+		Use:   "approve CSR",
+		Short: "approve a CSR.",
+		Long: `
+kubectl certificate approve allows a cluster admin to approve a certificate
```
Long should have a title line.

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,174 @@
+	options := CertificateOptions{}
+	cmd := &cobra.Command{
+		Use:   "approve CSR",
+		Short: "approve a CSR.",
```
Short should be a sentence (capitalized) but with no punctuation.

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,174 @@
+	return cmd
+}
+
+func RunCertificateApprove(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *CertificateOptions) error {
```
Probably should follow the kubectl options complete/verify/run convention as described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#command-implementation-conventions

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,174 @@
+		csr.Status.Conditions = append(csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
+			Type:           certificates.CertificateApproved,
+			Reason:         "KubectlApprove",
+			Message:        "This csr was approved by kubectl alpha approve",
```
Be sure to capitalize CSR correctly in all user facing messages and descriptions.

mikedanese commented:
@k8s-bot cvm gke e2e test this, issue: #35913

mikedanese commented:
@smarterclayton thanks for the doc link. I've addressed comments.

mikedanese commented:
@k8s-bot node e2e test this, issue: #35935

mikedanese commented:
@kubernetes/kubectl @pwittrock @smarterclayton can i get a final review? :+1:

smarterclayton commented on pkg/kubectl/cmd/certificates.go:
```diff
@@ -0,0 +1,197 @@
+		SECURITY NOTICE: Depending on the requested attributes, the issued certificate
+		can potentially grant a requester access to cluster resources or to authenticate
+		as a requested identity. Before approving a CSR, ensure you understand what the
+		signed certificate can do. This can be very useful but the security reprecussion
```
You can drop the last sentence here (to get rid of the typo on repercussion) - reads just as well without it.

smarterclayton commented:
One comment on description then LGTM (feel free to self apply)

saad-ali commented:
Removing 1.5 milestone. Looks like this missed the cut off for code freeze. Is this fixing a bug?

saad-ali commented:
> Removing 1.5 milestone. Looks like this missed the cut off for code freeze. Is this fixing a bug?

Looks like the original comment LGTM was before code freeze:

> smarterclayton commented 4 days ago
> One comment on description then LGTM (feel free to self apply)

Reinstating 1.5 label. Ok for post-freeze merge.

mikedanese commented:
@k8s-bot gce etcd3 e2e test this

bgrant0607 commented:
@mikedanese Would you be kind enough to regenerate these files? They contain the old warning boilerplate rather than the new simple auto-generated message.
